src/checkpoint/serializers/reference.h

```diff
@@ -0,0 +1,178 @@
+// Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_CHECKPOINT_SERIALIZERS_REFERENCE_H
+#define INCLUDED_CHECKPOINT_SERIALIZERS_REFERENCE_H
+
+#include "checkpoint/common.h"
+#include "checkpoint/serializers/base_serializer.h"
+#include "checkpoint/traits/user_traits.h"
+
+#include <type_traits>
+#include <cstdlib>
+#include <memory>
+
+namespace checkpoint {
+
+namespace {
+  template<typename T, typename... U>
+  struct first_type_matches : public std::false_type {};
+  template<typename T, typename... U>
+  struct first_type_matches<T, T, U...> : public std::true_type {};
+}
+
+template<typename SerT, typename... UserTraits>
+struct SerializerRef
+{
+  using TraitHolder = typename UserTraitHolder<UserTraits...>::BaseTraits;
+
+  //Passing initialization to the serializer implementation
+  template<typename... Args>
+  explicit SerializerRef(Args&&... args)
+  {
+    if constexpr(first_type_matches<std::shared_ptr<SerT>, std::decay_t<Args>...>::value and sizeof...(Args) == 1) {
+        //If only 1 arg which is a shared pointer to the base serializer, use that instead of initializing a new serializer.
+        impl = {std::forward<Args>(args)...};
+    } else {
+        impl = std::make_shared<SerT>(std::forward<Args>(args)...);
+    }
+  }
```

Comment:
I think this should be two overloads:
```cpp
explicit SerializerRef(std::shared_ptr<SerT> ptr);

template<typename... Args>
explicit SerializerRef(std::in_place_t, Args&&... args);
```

Using the `std::in_place` to avoid ambiguity.